packages/transformer/src/IModelTransformer.ts

```diff
@@ -2889,145 +2950,212 @@ export class IModelTransformer extends IModelExportHandler {
-          alreadyImportedElementInserts,
-          alreadyImportedModelInserts
-        );
+        if (relationshipECClassIds.has(ecClassId)) {
+          if (change.$meta?.classFullName === undefined) {
+            Logger.logError(
+              loggerCategory,
+              "ClassFullName was not found for relationship when reading changes. Relationship delete will not propagate.",
+              { relationshipId: change.ECInstanceId, ecClassId }
+            );
+            continue;
+          }
+          if (
+            change.SourceECInstanceId === undefined ||
+            change.TargetECInstanceId === undefined
+          ) {
+            Logger.logError(
+              loggerCategory,
+              "SourceECInstanceId or TargetECInstanceId was not found for relationship when reading changes. Relationship delete will not propagate.",
+              {
+                relationshipId: change.ECInstanceId,
+                ecClassId,
+                classFullName: change.$meta.classFullName,
+              }
+            );
+            continue;
+          }
+          await this.processRelationshipDeleteOp(
+            change.ECInstanceId,
+            change.$meta.classFullName,
+            change.SourceECInstanceId,
+            change.TargetECInstanceId
+          );
+        } else {
+          await this.processElementDeleteOp(
+            change.ECInstanceId,
+            alreadyImportedElementInserts,
+            alreadyImportedModelInserts,
+            elemIdToScopeEsa,
+            change.FederationGuid
+          );
+        }
+        deleteIdsProcessed.add(change.ECInstanceId);
       }
 
       csReader.close();
     }
+
+    await this.handleCustomChanges(hasElementChangedCache, deleteIdsProcessed);
+
     this._hasElementChangedCache = hasElementChangedCache;
     return;
   }
+
+  /**
+   * Helper function for processChangesets.
+   * Populates the '_deletedSourceRelationshipData' map, whose key is the id of the relationship in the source and the value is an object used to find that relationship in the target.
+   * @param changedInstanceId The id of the relationship that was deleted
+   * @param classFullName classFullName of relationship
+   * @param sourceIdOfRelationshipInSource the element Id acting as the source of the relationship in the sourceDb
+   * @param targetIdOfRelationshipInSource the element Id acting as the target of the relationship in the sourceDb
+   * @returns
+   */
+  private async processRelationshipDeleteOp(
+    changedInstanceId: Id64String,
+    classFullName: string,
+    sourceIdOfRelationshipInSource: Id64String,
+    targetIdOfRelationshipInSource: Id64String
+  ) {
+    // we need a connected iModel with changes to remap elements with deletions
+    const notConnectedModel = this.sourceDb.iTwinId === undefined;
+    const noChanges =
+      this.synchronizationVersion.index === this.sourceDb.changeset.index &&
+      this.exporter.sourceDbChanges?.isEmpty;
```

Comment:
```
 const noChanges =
      this.synchronizationVersion.index === this.sourceDb.changeset.index &&
      this.exporter.sourceDbChanges?.isEmpty;
```
`noChanges` will be always `false` when this.exporter.sourceDbChanges is `undefined`. 

Suuggest this: https://github.com/iTwin/imodel-transformer/pull/172/files#r1910400545

Reply:
Gah I am getting tripped up by so many of these conditional operators.

I think your suggestion is correct but how do you feel about the more explicit version:
```
    const noChanges =
      this.synchronizationVersion.index === this.sourceDb.changeset.index &&
      (this.exporter.sourceDbChanges === undefined || !this.exporter.sourceDbChanges.hasChanges);
```

Instead of `!this.exporter.sourceDbChanges?.hasChanges`

Reply:
Seems ok.